processor/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,504 @@
+					traceID := span.TraceID().HexString()
+					key := buildEdgeKey(traceID, span.SpanID().HexString())
+					isNew, err = p.store.UpsertEdge(key, func(e *store.Edge) {
+						e.TraceID = traceID
+						e.ConnectionType = connectionType
+						e.ClientService = serviceName
+						e.ClientLatencySec = float64(span.EndTimestamp()-span.StartTimestamp()) / float64(time.Millisecond.Nanoseconds())
+						e.Failed = e.Failed || span.Status().Code() == 2
+						p.upsertDimensions(e.Dimensions, rAttributes, span.Attributes())
+
+						// A database request will only have one span, we don't wait for the server
+						// span but just copy details from the client span
+						if dbName, ok := findAttributeValue("db.name", rAttributes, span.Attributes()); ok {
+							e.ConnectionType = store.Database
+							e.ServerService = dbName
+							e.ServerLatencySec = float64(span.EndTimestamp()-span.StartTimestamp()) / float64(time.Millisecond.Nanoseconds())
+						}
+					})
+				case ptrace.SpanKindConsumer:
+					// override connection type and continue processing as span kind server
+					connectionType = store.MessagingSystem
+					fallthrough
+				case ptrace.SpanKindServer:
+					traceID := span.TraceID().HexString()
+					key := buildEdgeKey(traceID, span.ParentSpanID().HexString())
+					isNew, err = p.store.UpsertEdge(key, func(e *store.Edge) {
+						e.TraceID = traceID
+						e.ConnectionType = connectionType
+						e.ServerService = serviceName
+						e.ServerLatencySec = float64(span.EndTimestamp()-span.StartTimestamp()) / float64(time.Millisecond.Nanoseconds())
+						e.Failed = e.Failed || span.Status().Code() == 2
+						p.upsertDimensions(e.Dimensions, rAttributes, span.Attributes())
+					})
+				default:
+					// this span is not part of an edge
+					continue
+				}
+
+				if errors.Is(err, store.ErrTooManyItems) {
+					totalDroppedSpans++
+					stats.Record(ctx, statDroppedSpans.M(1))
+					continue
+				}
+
+				// UpsertEdge will only return ErrTooManyItems
+				if err != nil {
+					return err
+				}
+
+				if isNew {
+					stats.Record(ctx, statTotalEdges.M(1))
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (p *processor) upsertDimensions(m map[string]string, resourceAttr pcommon.Map, spanAttr pcommon.Map) {
+	for _, dim := range p.config.Dimensions {
+		if v, ok := findAttributeValue(dim, resourceAttr, spanAttr); ok {
+			m[dim] = v
+		}
+	}
+}
+
+func (p *processor) onComplete(e *store.Edge) {
+	p.logger.Debug("edge completed")
+	p.aggregateMetricsForEdge(e)
+}
+
+func (p *processor) onExpire(*store.Edge) {
+	p.logger.Debug("edge expired")
+	stats.Record(context.Background(), statExpiredEdges.M(1))
+}
+
+func (p *processor) aggregateMetricsForEdge(e *store.Edge) {
+	metricKey := p.buildMetricKey(e.ClientService, e.ServerService, string(e.ConnectionType), e.Dimensions)
+	dimensions := buildDimensions(e)
+
+	// TODO: Consider configuring server or client latency
+	duration := e.ServerLatencySec
+
+	p.seriesMutex.Lock()
```

Comment:
If you don't need them to be serially updated, you can consider using an atomic integer instead of a lock + ints.

Reply:
I reviewed the implementation again, they need to be updated all at the same time, b/c `keyToMetric` can be modified in between.